.gitlab-ci.yml

```diff
@@ -1,259 +1,361 @@
+# resource-intensive jobs.  Besides running checks, this condenses the initial
+# cache generation into a single job for the later stages.  The first stage,
+# in particular, would otherwise suffer a heavy cache-miss penalty as its
+# jobs kick off in parallel.
+sanity_checks:
+  stage: 'sanity'
+  tags: ['docker']
+  script:
+    - 'bundle exec rake check:dot_underscore'
+    - 'bundle exec rake check:test_file'
+    - 'bundle exec rake pkg:check_version'
+    - 'bundle exec rake pkg:compare_latest_tag'
 
+tag_check:
+  stage: 'sanity'
+  only: ['tags']
+  tags: ['docker']
+  script: '[ "$CI_COMMIT_TAG" = "$MODULE_VERSION" ] || ( echo "ERROR: Tag does not match metadata version" && exit 1 )'
 
-# Keep an eye on the latest puppet 5
-# ----------------------------------
-pup5_latest-validation:
-  stage: validation
-  tags:
-    - docker
-  image: ruby:2.4
-  variables:
-    PUPPET_VERSION: '~> 5.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *validation_checks
-  allow_failure: true
 
-pup5_latest-unit:
-  stage: unit
-  tags:
-    - docker
-  image: ruby:2.4
-  variables:
-    PUPPET_VERSION: '~> 5.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
-  allow_failure: true
+#=======================================================================
+# Lint Tests
+#
 
+# Linting, for the most part, isn't affected by version changes in Puppet,
+# so we only test against the latest version for each MAJOR release.
+pup4_latest-lint:
+  <<: *lint_base
+  <<: *pup4_latest
+  image: 'ruby:2.1'
+
+pup5_latest-lint:
+  <<: *lint_base
+  <<: *pup5_latest
+  image: 'ruby:2.4'
+
+
+#=======================================================================
+# Unit Test Matrix
+#
+
+pup4_7-unit:
+  <<: *unit_base
+  <<: *pe_2016_4
+  image: 'ruby:2.1'
+
+pup4_8-unit:
+  <<: *unit_base
+  <<: *simp_6_0
+  image: 'ruby:2.1'
+
+pup4_10-unit:
+  <<: *unit_base
+  <<: *simp_6_1
+  image: 'ruby:2.1'
+
+pup5_3-unit:
+  <<: *unit_base
+  <<: *pe_2017_3
+  image: 'ruby:2.4'
+
+pup5_5-unit:
+  <<: *unit_base
+  <<: *pe_2018_1
+  image: 'ruby:2.4'
+
+# Use this special job to test all the latest stuff and update the cache.
+pup5_latest-unit:
+  <<: *unit_base
+  <<: *pup5_latest
+  image: 'ruby:2.4'
+  script:
+    - 'bundle update'
+    - 'bundle exec rake spec'
+  cache:
+    policy: 'pull-push'
 
 
-# Acceptance tests
 # ==============================================================================
-default:
-  stage: acceptance
-  tags:
-    - beaker
-  <<: *cache_bundler
-  <<: *setup_bundler_env
+# Acceptance tests
+#
+
+# Because acceptance tests are so much more expensive than other tests, this
+# test matrix is even more limited.  Here we stick to versions supported
+# by non-EOL meta-releases of SIMP and LTS versions of Puppet Enterprise.
+
+# ----------------------------------------------------------------------
+# Puppet 4.8 for SIMP 6.0 and PE 2016.5
+el-pup4_8-default:
```

Comment:
I suggest putting the version of puppet ahead at the front of job name.  There is limited space on the pipeline UI, and the puppet version is the most significant change.

The `el` is unnecessary; all of these tests are `el`